src/compiler/checker.ts

```diff
@@ -18999,6 +19007,49 @@ namespace ts {
+            const reactRefErr = diagnostics && compilerOptions.jsx === JsxEmit.React ? Diagnostics.Cannot_find_name_0 : undefined;
+            const reactNamespace = getJsxNamespace(node);
+            const reactLocation = isJsxOpeningLikeElement(node) ? node.tagName : node;
+            const reactSym = resolveName(reactLocation, reactNamespace, SymbolFlags.Value, reactRefErr, reactNamespace, /*isUse*/ true);
+            const factoryEntity = getJsxFactoryEntity(reactLocation);
+            if (!factoryEntity) {
+                return getJsxElementTypeAt(node) || errorType;
+            }
+            const factorySymbol = resolveEntityName(factoryEntity, SymbolFlags.Value, /*ignoreErrors*/ true, /*dontUseResolveAlias*/ false, reactLocation);
+            if (reactSym && factorySymbol && factorySymbol !== unknownSymbol) {
+                // Mark local symbol as referenced here because it might not have been marked
+                // if jsx emit was not react as there wont be error being emitted
+                reactSym.isReferenced = SymbolFlags.All;
+
+                // If react symbol is alias, mark it as referenced
+                if (reactSym.flags & SymbolFlags.Alias && !isConstEnumOrConstEnumOnlyModule(resolveAlias(reactSym))) {
+                    markAliasSymbolAsReferenced(reactSym);
+                }
+                const links = getNodeLinks(node);
+                if (!links.jsxFactoryCall) {
+                    const factoryExpression = createSyntheticExpression(node, getTypeOfSymbol(factorySymbol));
+                    const children = isJsxOpeningElement(node) ? node.parent.children : emptyArray;
+                    links.jsxFactoryCall = createCall(factoryExpression, /*typeArguments*/ undefined, [
+                        isJsxOpeningFragment(node)
+                            ? createSyntheticExpression(node, reactSym.exports ? getTypeOfSymbol(getSymbol(getExportsOfSymbol(reactSym), "Fragment" as __String, SymbolFlags.Value) || unknownSymbol) : emptyObjectType)
+                            : isJsxIntrinsicIdentifier(node.tagName) ? createSyntheticExpression(node.tagName, getLiteralType(idText((node.tagName as Identifier)))) : node.tagName,
+                        isJsxOpeningFragment(node) ? createSyntheticExpression(node, nullType) : createSyntheticExpression(node.attributes, checkMode => checkExpression(node.attributes, checkMode)),
+                        ...mapDefined(children, c => isJsxText(c) && c.containsOnlyWhiteSpaces ? undefined : createSyntheticExpression(c, isJsxText(c) ? stringType : (checkMode => checkExpression(c, checkMode))))
+                    ]);
+                    links.jsxFactoryCall.pos = node.pos;
+                    links.jsxFactoryCall.end = node.end;
+                    links.jsxFactoryCall.parent = node.parent;
+                }
+                const result = getReturnTypeOfSignature(getResolvedSignature(links.jsxFactoryCall));

```

Comment:
Could this potentially let us get rid of `JSX.Element` (or, heck, the entire `JSX` namespace)? 🤔 

If we can do that _and_ redefine `createElement` / `React.ReactElement` we might be able to get strong type checking for element children, and more 🎉

~~sounds like an excellent opportunity for a breaking change~~

Reply:
That would be what one would hope for, yeah. Technically we don't want to be rid of it quite yet because there's no way to define the factory function that correctly resolves overloads on components (or generics), but by and large if we had those this wouldn't need the namespace at all.